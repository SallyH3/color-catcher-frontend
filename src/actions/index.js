export const makePalette = (palette) => {
	return {
		type: 'MAKE_PALETTE',
		palette
	}
}

export const clearPalette = (palette) => {
	return {
		type: 'CLEAR_PALETTE',
		palette
	}
}