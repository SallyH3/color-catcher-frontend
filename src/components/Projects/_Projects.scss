.project-holder {
	width: 27%;
	min-height: 200px;
	height: 100%;
	background-color: hsl(0, 77%, 71%);
}

.name-proj-input {
	border: 1px solid hsl(0, 0%, 0%);
	border-radius: 5px;
	height: 25px;
	margin-top: 10px;
	margin: 5px;
}

.create-btn {
	border: 1px solid hsl(0, 0%, 0%);
	border-radius: 3px;
	cursor: pointer;
	height: 25px;
	width: 60px;
}

.project-form,
.projects-title,
.project-name {
	margin: 10px;
}

.project-name {
	font-size: 16px;
	letter-spacing: 1px;
}

.project-delete-btn {
	border: 1px solid hsl(0, 0%, 0%);
	margin: 10px;
	border-radius: 5px;
	height: 25px;
	width: 100px;
	cursor: pointer;
}