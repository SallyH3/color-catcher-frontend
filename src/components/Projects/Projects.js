import React, { Component } from 'react';
import './_Projects.scss';
import { connect } from 'react-redux';
import { setProjects, deleteProject, addProject, pickProject, setPalettes } from '../../actions';

export class Projects extends Component {
	constructor() {
		super();
		this.state = {
			title: '',
		}
	}

	componentDidMount = () => {
		fetch('http://localhost:3001/api/v1/project')
		.then(response => response.json())
		.then(projects => this.props.setProjects(projects))

		fetch('http://localhost:3001/api/v1/palettes')
		.then(response => response.json())
		.then(palettes => this.props.setPalettes(palettes))
	}
	
	handleChange = (e) => {
		this.setState({title: e.target.value});  
  }

	makeProject = (e) => {
		e.preventDefault();
		this.postProject(this.state.title);
		this.setState({title: ''});
	}

	postProject = (title) => {
		let name = title
		let options = {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        name: name
      })
    };

		fetch('http://localhost:3001/api/v1/project', options)
		.then(response => {
			if(!response.ok) {
				throw Error('Error posting project')
			} else {
				return response
			}
		})
		.then(result => result.json())
		.then(body => this.props.addProject({id: body.id, name}))
	}

	deleteProject = (e) => {
		let options = {
      method: "DELETE"
    };
		const idToDelete = e.target.parentElement.getAttribute('data-key')
		this.props.deleteProject(idToDelete)
		fetch(`http://localhost:3001/api/v1/project/${idToDelete}`, options)
	}

	chooseProject = (e) => {
		const id = e.target.parentElement.getAttribute('data-key')
		const project = this.props.projects.find(project => {
			return project.id === Number(id)
		})
		this.props.pickProject(project)
	}

	render() {
		const projectNames = this.props.projects.length ? this.props.projects.map(project => {
			return (
				<div data-key={project.id} key={project.id}>
					<h2 onClick={this.chooseProject}>{project.name}</h2>
					<button onClick={this.deleteProject}>
						Delete
					</button>
				</div> 
			)
		}) : ''

		return (
			<div className='project-holder'>
				<form> 
					<label>Name Project</label>
					<input
						type="text"
	          onChange={this.handleChange}
	          name='title'
	          value={this.state.title}
	         />
					<button onClick={this.makeProject}>Create</button> 
				</form>
				<h1>Projects</h1>
				{projectNames}
			</div>
		)
	}
}

export const mapStateToProps = (state) => ({
	projects: state.projects,
	project: state.project
});

export const mapDispatchToProps = (dispatch) => ({
	setProjects: (projects) => dispatch(setProjects(projects)),
	addProject: (project) => dispatch(addProject(project)),
	pickProject: (project) => dispatch(pickProject(project)),
	deleteProject: (id) => dispatch(deleteProject(id)),
	setPalettes: (palettes) => dispatch(setPalettes(palettes))
});

export default connect(mapStateToProps, mapDispatchToProps)(Projects);