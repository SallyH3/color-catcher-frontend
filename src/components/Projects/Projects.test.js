import React from 'react';
import { shallow } from 'enzyme';
import { Projects, mapStateToProps, mapDispatchToProps } from './Projects'
import { setProjects, deleteProject, addProject } from '../../actions';

describe('Projects', () => {
	let wrapper;
	let mockSetProjects;
	let mockDeleteProject;
	let mockAddProject;
	let mockProjects;

	beforeEach(() => {
		mockAddProject = jest.fn();
		mockDeleteProject = jest.fn();
		mockSetProjects = jest.fn();
		mockProjects = [{
			name: 'project one',
			id: 1
		}]

		wrapper = shallow(
			<Projects 
				setProjects={mockSetProjects}
				addProject={mockAddProject}
				deleteProject={mockDeleteProject}
				projects={mockProjects}
			/>
		);
	});

	it('should match the snapshot', () => {
		expect(wrapper).toMatchSnapshot();
	});

	describe('handleChange', () => {
		it('should set state with title', () => {

			const mockTitleEvent = {
				target: {
					value: 'Project One'
				}
			};

			const wrapper = shallow(
				<Projects 
					projects={mockProjects}
				/>
			);

			wrapper.instance().handleChange(mockTitleEvent);
			expect(wrapper.state().title).toEqual('Project One');
		});
	});

	describe('mapStateToProps', () => {
		it('should return an array of objects', () => {
			const mockState = {
			palette: [
				{	locked: false,
					color: "#909C0E"
				},
				{	locked: false,
					color: "#82BBC8"
				},
				{	locked: false,
					color: "#82BBC8"
				},
				{
					locked: false,
					color: "#82BBC8"
				},
				{	locked: false,
					color: "#909C0E"
				}
			],
				project: [
					{	name: 'project one',
						id: 1
					}
				]
			};

			const expected = { 
				projects: [
				{	name: 'project one',
					id: 1
				}
			]
		};
		
		const mappedProps = mapStateToProps(mockState);

			expect(mappedProps).toEqual(expected)
		});
	});

	describe('mapDispatchToProps', () => {
		it('should call dispatch when using setProjects', () => {

			const mockProjects = [
				{
					name: 'project one',
					id: 1
				},
				{
					name: 'project two',
					id: 2
				}
			]

			const mockDispatch = jest.fn()

			const actionToDispatch = setProjects(mockProjects)

			const mappedProps = mapDispatchToProps(mockDispatch);

			mappedProps.setProjects(mockProjects)

			expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)

		});

		it('should call dispatch when using addProject', () => {

			const mockProject = {
				name: 'project one',
				id: 1
			};

			const mockDispatch =jest.fn();

			const actionToDispatch = addProject(mockProject);

			const mappedProps = mapDispatchToProps(mockDispatch);

			mappedProps.addProject(mockProject);

			expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
		});

		it('should call dispatch when using deleteProject', () => {

			const mockId = 1;

			const mockDispatch = jest.fn();

			const actionToDispatch = deleteProject(mockId);

			const mappedProps = mapDispatchToProps(mockDispatch);

			mappedProps.deleteProject(mockId);

			expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
		});
	});
});


