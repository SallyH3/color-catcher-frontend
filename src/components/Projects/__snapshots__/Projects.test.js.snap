// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Projects should match the snapshot 1`] = `
<div
  className="project-holder"
>
  <form
    className="project-form"
  >
    <label>
      Name Your Project
    </label>
    <input
      className="name-proj-input"
      name="title"
      onChange={[Function]}
      type="text"
      value=""
    />
    <button
      className="create-btn"
      onClick={[Function]}
    >
      Create
    </button>
  </form>
  <h1
    className="projects-title"
  >
    Projects
  </h1>
  <div
    data-key={1}
    key="1"
  >
    <h2
      className="project-name"
      onClick={[Function]}
    >
      project one
    </h2>
    <button
      className="project-delete-btn"
      onClick={[Function]}
    >
      Delete Project
    </button>
  </div>
</div>
`;
